doc/decisions/0_drafts/transformations.md

```diff
@@ -0,0 +1,157 @@
+3.
+
+## Assumptions
+
+1. False positives for change tracking algorithms are only a minor problem.
+   False positives are that changes are detected, even though nothing changed.
+2. There is no reason to modify or delete existing `meta:/` keys.
+3. Newly generated `meta:/...` keys can
+   - either stay and get permanently stored during `kdbSet`
+   - or be written as `meta:/generated/...`.
+     The `meta:/generated/...` metakeys are never stored and are automatically removed during `kdbSet` before `storage` is called.
+
+## Considered Alternatives
+
+### Enable setting and unsetting of the `keyNeedsSync` flag
+
+Require plugins that rename keys to remove the `keyNeedsSync` flag.
+This would require making the 'key sync' APIs public.
+Letting user code modify these internal flags may lead to serious bugs.
+This does not solve the problem of changing values.
+
+### Use change tracking to replace the `keyNeedsSync` flag
+
+Eliminate the need of the `keyNeedsSync` flag by utilizing the planned change tracking API.
+This does not solve the problem of changing values and false positives.
+
+### Store original names and values in metadata.
+
+For key name transformations require that transformation plugins set a metakey (e.g. `meta:/elektra/runtimename`) with the runtime name before they do any transformations in the `kdbSet` phase.
+This _must not_ be done if this metakey already exists, i.e. another plugin already tranformed it beforehand.
+This allows notification and change tracking functionality to work determine and work with the runtime name of the key.
+
+A similiar thing was already attempted for values, i.e. `meta:/origvalue`.
+
+### Create an API for transformations
+
+Plugins must use a special function to transform key names, e.g.:
+
+```c
+typedef const char * (*ElektraNameTransform)(const Key *);
+
+int elektraApplyNameTransform (KeySet * ks, ElektraNameTransform transform);
+```
+
+How the `elektraApplyNameTransform` function marks the original name is an internal implementation detail.
+May be as `meta:/` or something else entireley.
+
+Something similar could be done for the value of a key as well.
+
+### Introduce a new phase for transformations
+
+We could also introduce a new phase between before/after storage exclusively for transformations.
+Then we can just do a "fake" call to that phase to get back the transient names for change tracking.
+
+### Call value transformation plugin(s) when `keySetValue` / `keySetString` is called
+
+After a value has been set by the user, call the transformation plugin.
+We could store those callbacks as metakeys, i.e. `meta:/generated/transformation/value/callback/#1`.
+
+In `keySetRaw` we then call all those callbacks.
+We also need to provide a simple API to plugins to register such callbacks for a key.
+
+As the transformations will be applied before `kdbSet`, this will elimate lots of possible false positives in changetracking.
```

Comment:
1. This should be done for name and value, not just value to be consistent.
2. I don't like that it makes `libelektra-core` more complex.
3. The name `meta:/generated/transformation/value/callback/#1` is very long. Using metakeys with such long names uses a lot of memory compare to the naive solution of adding a linked list 
    ```c
    struct callback {
        callback_fn function;
        struct callback * next;
    }
    ```
    to the key struct.
4. It's a neat idea, but doesn't work. There is a very simple case that isn't covered:
    ```c
    kdbGet (kdb, ks, parent);
    ksAppendKey (ks, keyNew ("user:/foo", KEY_VALUE, "new value", KEY_END);
    kdbSet (kdb, ks, parent);
    ```
    The new key doesn't have the callback and therefore isn't transformed before `kdbSet`.

Reply:
Good points, but

1. Does this actually make sense for names? IMHO names should be handled as close to the storage plugins as possible. But yes, nothing is preventing us from doing it for names also.
2. This is true, but we already do similar things. i.e. `meta:/binary`.
3. Yes, true. How we implement it is not very important to me.
4. It works to eliminate false positives. 
   - A new key is always a change so that it would get detected correctly. 
   - The transformation plugins could be required to handle those cases too, e.g. if no callback is on that key, do the transformation in `kdbSet` as it is now.

Reply:
Ad 4: What exactly would the callback do?

To me it sounded like the callback would perform the transformation. But that would mean that a newly added key is never transformed, unless the plugin also checks during `kdbSet`. But then we're back to the same old problem.

To use a concrete example, take the `user:/limits/openfiles` example you gave in the "Problem" section.

How would this solution work in this snippet?

```c
kdbGet (kdb, ks, parent);
ksAppendKey (ks, keyNew ("user:/limits/openfiles", KEY_VALUE, "23", KEY_END);
kdbSet (kdb, ks, parent);
```

IMO this is not solved at all. There is no callback, because it is a new key and if the plugin "simply runs"  during `kdbSet`, we haven't solved anything compare to the problem statement.

Reply:
Yes, there would be 2 different possibilities where transformations happen:
1. For keys returned by `kdbGet`, i.e. with the callback set --> in the callback
2. new keys, i.e. keys that do not have a callback --> in `kdbSet`

In the case of the example, the transformation on `user:/limits/openfiles` would be in `kdbSet`.

As this is a new key, it doesn't matter that the transformation is done later for notification and changetracking plugins. They should detect this key as modified anyway.

But for existing keys, the transformation would take part long before `kdbSet`, so here we can avoid false positives.

Reply:
> They should detect this key as modified anyway.

But wasn't that the point of your example? The key value wouldn't actually be modified, because the plugin changes the value back to `1`.

Reply:
In the case of a new key, there wouldn't be a previous value?

Reply:
If you are hinting at the case, that `kdbGet` already returns a key `user:/limits/openfiles` and we are overriding this key with `ksAppendKey` with a new key: then we also need to check if the overridden key has a callback and the new one does not. 

If this is the case, then copy the callback to the new case, and execute the callback. So this would be a 3rd possibility where the transformation happens.

Reply:
> copy the callback to the new key, and execute the callback

So `ksAppendKey` would have to do that?

This solution is getting really complicated with a lot of moving parts...

Reply:
> So ksAppendKey would have to do that?

Yes, both `ksAppendKey ` and `ksAppend`.

> This solution is getting really complicated with a lot of moving parts...

I don't think it's _that_ complicated, but sure. However, this is the first solution that would really eliminate false positives. I'm always open to new solutions for this problem.

Reply:
I updated the decision to include what has been discussed here.